articles/qos_deadline_liveliness_lifespan.md

```diff
@@ -25,102 +25,155 @@ Glossary:
+Services will support the following levels of liveliness.
+- LIVELINESS_DEFAULT - Use the ROS specified default for liveliness (which is LIVELINESS_AUTOMATIC).
+- LIVELINESS_AUTOMATIC - The signal that establishes a Service Owner is alive comes from the ROS rmw layer.
+- LIVELINESS_MANUAL_NODE - The signal that establishes a Service is alive is at the node level. A message on any outgoing channel on the node or an explicit signal from the application to assert liveliness on the node will mark all Services on the node as being alive.
+- LIVELINESS_MANUAL_SERVICE - The signal that establishes a Service is alive is at the Service level. Only sending a response on the Service or an explicit signal from the application to assert liveliness on the Service will mark the Service as being alive.
 
-Some of the primary DDS vendors that ROS supports already provide support for these QoS settings.
-However, the default vendor, Fast-RTPS, does not yet provide full support for these new QoS settings in their library.
-This does not have to be a blocker for adding support for these policy settings to ROS.
-We can start adding support for them and just log warning messages when a user attempts to specify an unsupported QoS setting.
-When support is eventually added for the policies in the vendor library then we can enable it in the rmw implementation.
+In order for a Subscriber to listen to a Publisher's Topic the liveliness they request must be greater than the liveliness set by the Publisher.
 
-### Native ROS Support vs Vendor Configuration
+Service Owners and Clients will each specify two liveliness policies, one for the liveliness policy pertaining to the Owner and one pertaining to the Client.
+In order for a Client to connect to an Owner to make a request the Client_Liveliness level requested by the Owner must be greater than the level provided by the Client and the Owner_Liveliness requested by the Client must be greater than the level provided by the Owner.
+
+### Lifespan
 
-It may be argued that many DDS QoS settings could be used with ROS by simply specifying them using the vendor specific configuration mechanism, such as the xml configuration files used by RTI Connext.
-This may be acceptable for some QoS policies that only impact the behavior of the underlying DDS middleware, but in the case where support for QoS policy impacts the behavior of the application code this is not good enough.
+The lifespan policy establishes a contract for how long a message remains valid.
+For Topic Subscribers it establishes the length of time a message is considered valid, after which time it will not be received.
+For Topic Publishers it establishes the length of time a message is considered valid, after which time it will be removed from the Topic history and no longer sent to Subscribers.
+For Service Owners it establishes the length of time a request is considered valid, after which time the Owner will not receive and process the request.
```

Comment:
Will clients be informed that their request has timed out?

Lifespan works well with publish-and-forget data, because the publisher is not the one who worries about whether a piece of data was received or not, but it is not as simple for requests.

Reply:
Good question. As far as I know, service clients don't really support timeouts in the rcl layer or below, correct me if I'm wrong. My thought hadn't been to use this to enforce a timeout mechanism, but rather a way to prevent stale requests from reaching a service provider. I can see three options to move forward here: 
1) Explicitly state that clients will not be notified of requests that are removed due to an expired lifespan. This still provides some benefits of service providers not receiving stale requests.
2) Add the concept of timeouts to service clients in the rcl/rmw layers to support a notification mechanism and cleanup resources. 
3) Remove lifespan support for Services from the design document for now. 

I'm leaning towards the third option as that provides the easiest and lowest risk path forward in terms of actually implementing the design. Any additional thoughts?

Reply:
3 is certainly the easiest; it doesn't preclude adding it back in later when we figure out how to do it properly.

I wonder if it would be usable for asynchronous services, where apart from asking the future for the result or blocking for the result, you could also ask the future if it is still possible to get a result, i.e. if the service is still active.

Reply:
Cool, I'm going to update the doc to make 3 plan of record for now. 

Seems like it should definitely be possible for asynch requests to be able to resolve the future with a failure code if the executor is given a way to check on the status of the request. I'd need to dig more into the how service futures are resolved to understand the implications.